src/java/org/apache/cassandra/security/DefaultSslContextFactoryImpl.java

```diff
@@ -0,0 +1,369 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.security;
+
+import java.io.File;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.handler.ssl.CipherSuiteFilter;
+import io.netty.handler.ssl.ClientAuth;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslProvider;
+
+/**
+ * Cassandra's default implementation class for the configuration key {@code ssl_context_factory}. It uses
+ * file based keystores.
+ */
+public final class DefaultSslContextFactoryImpl implements ISslContextFactory
```

Comment:
I can see having this as `final` being important to know you're definitely getting an unmodified version of the class, but it means that somebody who wanted to just tweak a few settings has to deal with the maintenance issues of keeping their copy up to date.

What do you think about dropping the final from this class and adding a new `public final class DefaultSslContextFactory extends DefaulSslContextFactoryImpl`. If somebody just wanted to tweak some of the Netty builder settings, though they could just override the `buildKeyManagerFactory`/`buildTrustManagerFactory` and if `useOpenSsl` gets moved into this module as a something like `getSslProvider()`.

Reply:
This is good point, related to what I was thinking recently ... What if somebody takes the default implementation like this and it will put it into "custom class" config param in cassandra.yaml? (That also means that it would get all params + custom params into its constructor as side effect).

So I think it definitely makes sense to make this not final if there is ever somebody who wants to build on top of this for whatever reason. We should enable to eat our own dog food and extend what we did. This implementation, even the default one, should be treated just like all other ones. It just happens to be _default one_.

Reply:
I created 'AbstractSslContextFactory' and `final` DefaultSslContextFactory classes with Abstract class not dealing with any file based credentials.